### PR TITLE
Use `sc_network::NetworkStateInfo` instead of implementing redundant traits

### PR DESCRIPTION
**Pull request by kigawas**
ref #4439

### CODE REVIEW

File: client/authority-discovery/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
+use std::sync::{Arc, Mutex};
```
Comment (tomusdrw) [inline client/authority-discovery/src/tests.rs]: This needs preamble. Why move the tests in the first place though? It's a common convention to just have an inline module.

File: client/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -64,7 +64,7 @@ use error::{Error, Result};
 use log::{debug, error, log_enabled, warn};
 use libp2p::Multiaddr;
 use sc_network::specialization::NetworkSpecialization;
-use sc_network::{DhtEvent, ExHashT};
+use sc_network::{DhtEvent, ExHashT, NetworkStateInfo};
```
Comment (tomusdrw) [inline client/authority-discovery/src/lib.rs]: I feel like these things (trait + primitives) should rather be moved to `primitves`, but it's a separate discussion.

Comment (kigawas) [reply client/authority-discovery/src/lib.rs]: You can check #4450

File: client/authority-discovery/src/tests.rs (hunk 1 above)
Comment (kigawas) [reply client/authority-discovery/src/tests.rs]: Besides less cognitive load, it'll be easier for me to format the code with `rustfmt` in a separate file :)

**Comment by kigawas** [comment]
```
warning: unused variable: `message`
   --> primitives/runtime/src/lib.rs:383:3
    |
375 |       /// A custom error in a module
    |  _____-
376 | |     Module {
377 | |         /// Module index, matching the metadata module index
378 | |         index: u8,
...   |
382 | |         #[codec(skip)]
383 | |         message: Option<&'static str>,
    | |         ^^^^^^-
    | |_______________|
    |                 help: try ignoring the field: `message: _`
    |
    = note: `#[warn(unused_variables)]` on by default
```

this made the build fail

File: client/authority-discovery/src/tests.rs [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+use sp_core::{testing::KeyStore, ExecutionContext, NativeOrEncoded};
+use sp_runtime::traits::Zero;
+use sp_runtime::traits::{ApiRef, Block as BlockT, NumberFor, ProvideRuntimeApi};
+use sp_test_primitives::Block;
+
+use super::*;
+
+#[test]
+fn interval_at_with_start_now() {
+	let start = Instant::now();
```
[the remark this answers is not in the document]
Comment (kigawas) [reply client/authority-discovery/src/tests.rs]: Can you elaborate?

File: client/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -96,10 +96,9 @@ const MAX_NUM_SENTRY_ADDRESSES_PER_AUTHORITY: usize = 5;
 pub struct AuthorityDiscovery<Client, Network, Block>
 where
 	Block: BlockT + 'static,
-	Network: NetworkProvider,
+	Network: NetworkProvider + NetworkStateInfo,
```
[the remark this answers is not in the document]
Comment (kigawas) [reply client/authority-discovery/src/lib.rs]: I'll do it. But actually we already have `put_value` and `get_value` in trait `Behaviour`, there's still some place to improve here. We can leave it out to other prs.

File: client/authority-discovery/src/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kigawas) [reply client/authority-discovery/src/tests.rs]: I just used rustfmt with hard tabs, it should has no much difference

File: client/network/src/service.rs [hunk 5]
```diff
@@ -422,11 +422,6 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> NetworkWorker
 }
 
 impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> NetworkService<B, S, H> {
-	/// Returns the network identity of the node.
-	pub fn local_peer_id(&self) -> PeerId {
```
[the remark this answers is not in the document]
Comment (kigawas) [reply client/network/src/service.rs]: We can even do better. I already addressed this in #4450

File: client/authority-discovery/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kigawas) [reply client/authority-discovery/src/lib.rs]: I'll do it. But actually we already have `put_value` and `get_value` in trait `Behaviour`, there's still some place to improve here.

**Comment by kigawas** [comment]
About the ci fix, since it's very simple and there should exist more specific solution to it, I'd like to get it merged in this pr and try fixing it in another way later.

The problem is rather subtle and urgent - it'll block everyone until it gets fixed.

**Comment by mxinden** [comment]
@kigawas in regards to the CI failures: https://github.com/paritytech/substrate/pull/4469 has been merged within the last 3 hours with a green CI, thus I think current `master` should be fine now.

File: client/authority-discovery/src/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomusdrw) [reply client/authority-discovery/src/tests.rs]: The first tab looks like 2 spaces on github diff, but the indentation is correct here.

[the remark this answers is not in the document]
Comment (mxinden) [reply client/authority-discovery/src/tests.rs]: Cool and sorry about that.

**Comment by tomusdrw** [comment]
Good catch @mxinden, we should not bundle those changes in that PR unless they are necessary. @kigawas could you please revert `Cargo.lock` changes and resubmit that as a separate PR so that we can easily review them in isolation?

**Comment by kigawas** [comment]
fixed

**Pull request merged**